File: backtracking/9_palindrome_partition.cpp
/* ************************************************************************
> File Name:     isPalindrome.cpp
> Author:        xmw01
> Created Time:  Thu 11 May 2023 09:38:06 AM CST
> Description:   
 ************************************************************************/

/*
给定一个字符串 s，将 s 分割成一些子串，使每个子串都是回文串。

返回 s 所有可能的分割方案。

示例: 输入: "aab" 输出: [ ["aa","b"], ["a","a","b"] ]
*/

#include<iostream>
#include<string>
#include<vector>

using namespace std;

class Solution{
private:
    vector<vector<string> > result;
    vector<string> path;
    void backtracking(string& s, int startIndex){
        //if the staring position is larger than s, a partiont scheme has been find
        if(startIndex >= s.size()){
            result.push_back(path);
            return;
        }
        for(int i = startIndex; i < s.size(); i++){
            //it's palindrom, next
            if(isPalindrome(s, startIndex, i)){
                string tmp = s.substr(startIndex, i - startIndex + 1);
                path.push_back(tmp);
            }else{
                //not palindrom, skip it
                continue;
            }
            backtracking(s, i + 1);   //backtracking
            path.pop_back();
        }
    }
public:
    bool isPalindrome(string& s, int begin, int end){
        for(int i = begin, j = end; i < j; i++, j--){
            if(s[i] != s[j]){
                return false;
            }
        }
        return true;
    }

    vector<vector<string> > palindrom_partion(string& s, int startIndex){
        cout << s << endl;
        result.clear();
        path.clear();
        backtracking(s, 0);
        return result;
    }
    
};

int main()
{
    Solution solution;
    string s ("aab");
    vector<vector<string> > output;
    output = solution.palindrom_partion(s, 0);
    cout << output.size() << endl;
    for (vector<vector<string> >::iterator iter = output.begin(); iter != output.end(); iter++){
        for(vector<string>::iterator j = iter->begin(); j != iter->end(); j++){
            // must use "->" for pointer
            cout << *j << " ";
        }
        cout << endl;
    }
}

File: backtracking/python/9_palindrome_partition.py
#!/mnt/lustre02/jiangsu/aispeech/home/xmw01/tools/anaconda3/bin/python
# -*- coding: UTF-8 -*-

from typing import List

"""
给定一个字符串 s，将 s 分割成一些子串，使每个子串都是回文串。

返回 s 所有可能的分割方案。

示例: 输入: "aab" 输出: [ ["aa","b"], ["a","a","b"] ]
"""

class Solution:
    def __init__(self) -> None:
        self.result = []
        self.path = []

    def palindromePatition(self, s: str) -> List[List[str]]:
        
        if (len(s) == 0):
            return
        self.result.clear()
        self.path.clear()

        self.backtracking(s, 0)
        return self.result

    def backtracking(self, s: str, startIndex: int) -> None:
        
        if (startIndex >= len(s)):
            self.result.append(self.path[:])
        
        for i in range(startIndex, len(s)):
            #注意不是 s[startIndex, i + 1]
            temp = s[startIndex : i + 1]
            if temp == temp[::-1]:
                self.path.append(temp)
                self.backtracking(s, i + 1)
                self.path.pop()
            else:
                continue

if __name__ == '__main__':
    s = "aab"
    
    solution = Solution()
    result = solution.palindromePatition(s)
    print(len(result))
    for item in result:
        for str_1 in item:
            print(str_1, end = ' ')
        print()